add deprecate guidelines doc [draft]

jchip opened the pull request.

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+
+## What consititutes an unmaintained package?
+
+- The author is no longer responding to questions, issues, PRs, or making any updates.
```
maybe worth qualifying here that they’re updates needed for particular reasons? if the only things the author is ignoring are feature requests the author doesn’t want, the package isn’t necessarily unmaintained.

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+- A different package that's more active and the author acknowledged as the replacement.
+- Critical issues exist for the package and not being addressed
+  - Known vulnerabilities identified by `npm audit` or other parties
+  - Package is known to fail for LTS NodeJS
```
```suggestion
  - Package is known to fail for LTS node.js
  - semver violations (any non-bug fix break in a non-major, eg)
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+
+- File issue in repo to encourage author to deprecate a version that:
+  - Has known critical bugs and should be avoided
+  - Known to fail for LTS NodeJS
```
```suggestion
  - Known to fail for LTS node.js
```
(not sure what NodeJS is :-p)

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+
+## Identify replacement
+
+- If a package is fully unmaintained, then a replacement should be identify and add to the deprecate message.
```
```suggestion
- If a package is fully unmaintained, then a replacement should be identified and added to the deprecation message.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+## Identify replacement
+
+- If a package is fully unmaintained, then a replacement should be identify and add to the deprecate message.
+- If no replacement exist, then should identify the safe versions to use in deprecate message.
```
```suggestion
- If no replacement exists, then should identify the safe versions to use in the deprecation message.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+  - but only owners can `npm deprecate` a package.
+  - if author simply can't be reached and package is very outdated, then need to contact npm to get access to deprecate package.
+
+* A cli to allow `npm deprecate` a range of versions on a package.
```
that’s already how `npm deprecate` works

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+  - if author simply can't be reached and package is very outdated, then need to contact npm to get access to deprecate package.
+
+* A cli to allow `npm deprecate` a range of versions on a package.
+* or `npm deprecate` versions with published date older than a given time
```
that seems an interesting feature to propose to npm, but it’s out of scope of this group

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+- file high priority issues in the package's repo
+
+  - if npm audit identified vulnerabilities that are critical
+  - if package is broken or fail to install for a LTS release of NodeJS
```
```suggestion
  - if fails to install/build for an LTS release of node.js
```
what constitutes “broken”?

Eomm commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,72 @@
+  - Has known critical bugs and should be avoided
+  - Known to fail for LTS Node.js
+  - Has known critical vulnerabilities
+
```
Could we add "Suggest to join the package-maintenance program"?
Some package could need help and doesn't know we are here to help them (with tools almost)

mhdawson commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -12,12 +12,12 @@ These guidelines exist to help with the course of actions to take in the event t
 ## What consititutes an unmaintained package?
 
-- The author is no longer responding to questions, issues, PRs, or making any updates.
+- The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there" inquries for critical issues.
```
`and specifically repeat "are you there" inquries for critical issues.`  maybe this should be

` and has not responded to repeated "are you there inquiries" on critical issues`

mhdawson commented:
I think this might be mixing different concerns.  What should a consumer do and what might this team do.  Would it make sense to break it out into those 2 sections or possibly 2 doc?

jchip commented:
I made some changes with target audience in mind.  Please re-review.

Eomm commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+If you are using a package that meets your requirements but you found issues or need enhancements, then these are some actions you can follow:
+
+- Note the deprecation messages during `npm install` and follow the instruction if the message offered any.
+- Check the package's npm page at `https://www.npmjs.com/package/<package-name>`. ie: <https://www.npmjs.com/package/express>
```
I would add also "Check the eventually `support` field in package.json" (refer to pr #139 )

Then I think this doc is clear enough ✌

dominykas commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -1,72 +1,107 @@
+There are some ongoing discussion related to npm to make managing unmaintained packages easier.
 
-## What user can do to avoid deprecate versions
+- Package override RFC https://github.com/aeschright/rfcs/blob/ca127d7c5fb7ecde534172d771351c0b8f819e79/accepted/0009-package-overrides.md
```
[npm@6.9.0](https://npm.community/t/release-npm-6-9-0/5911) allows package aliases - I have not yet looked into it in depth, and while the linked RFC here covers a method for override discovery, once a suitable replacement is identified - it might be worth mentioning the alias functionality?

dominykas commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -1,72 +1,107 @@
+### Taking Further Actions
 
-## `npm deprecate` command
+Once you've identified a package may be unmaintained, please open an issue for the Node.js package-maintenance working group to check and identify path forward.
```
Is this the right recommendation? I would like to see package-maintenance group to be the last resort (depending on popularity of the package, of course), but I would generally recommend looking for alternatives (are there any tools out there?) and then forking, i.e. dyi?

ljharb replied:
Forking is an important ability, but I see that as more of a last resort than asking this group - creating forks further fractures the ecosystem, and doesn't help all of the users who are still using the unmaintained version unwittingly.

jchip replied:
if no one can reach the author then there's really not much anyone could do.  I was thinking that the best this group could do is get npm ownership and actively advertise for someone interested in becoming the maintainer, and/or add deprecate message to identify alternative if one exist.

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+# How?
+
+These guidelines exist to help with the course of actions for package consumers and owners to take in the event that a package becomes unmaintained.
```
```suggestion
These guidelines exist to help with a course of action for package consumers and owners to take in the event that a package becomes unmaintained.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+## What consititutes an unmaintained package?
+
+It's not always obvious if a package is unmaintained. Many packages are small and considered completed as published. Long period of no activity does not mean unmaintain.
```
```suggestion
It's not always obvious if a package is unmaintained. Many packages are small and considered completed as published. A long period of no activity does not imply that a project is not maintained.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+It's not always obvious if a package is unmaintained. Many packages are small and considered completed as published. Long period of no activity does not mean unmaintain.
+
+Generally the question "is this package unmaintained?" arises when users found it to have issues that are not getting reponse or being addressed.
```
```suggestion
Generally the question "is this package unmaintained?" arises when users find it to have issues that are not getting responded to, or being addressed.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+Generally the question "is this package unmaintained?" arises when users found it to have issues that are not getting reponse or being addressed.
+
+Some criterias to determine if it's unmaintained:
```
```suggestion
Some criteria to determine if it's unmaintained:
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+Some criterias to determine if it's unmaintained:
+
+- The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there inquries" for critical issues.
```
```suggestion
- The author is no longer responding to questions, issues, or PRs; making any updates; and specifically, not responding to repeat "are you there" inquiries for critical issues.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+- The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there inquries" for critical issues.
+- The author may have explicitly indicated that they will stop all activities on the package.
+- A different package that's more active and the author acknowledged as the replacement.
```
```suggestion
- The author has explicitly indicated that an alternative package should be used instead.
```

mhdawson replied:
@jchip  can you accept this suggestion?

ljharb replied:
ping @jchip

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+- The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there inquries" for critical issues.
+- The author may have explicitly indicated that they will stop all activities on the package.
+- A different package that's more active and the author acknowledged as the replacement.
+- Critical issues exist for the package and not being addressed
```
```suggestion
- Critical issues exist for the package that are not being addressed
```

mhdawson replied:
@jchip can you accept this suggestion

ljharb replied:
ping here too

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+- Note the deprecation messages during `npm install` and follow the instruction if the message offered any.
+- Check the package's npm page at `https://www.npmjs.com/package/<package-name>`. ie: <https://www.npmjs.com/package/express>
+  - Check the "last publish" date and versions history to get a sense of its latest activies.
```
```suggestion
  - Check the "last publish" date and versions history to get a sense of its latest activity. Note that you may want to `npm show $package time` to see all the published versions and times, regardless of their dist-tag setting.
```

ljharb replied:
and here

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+- Note the deprecation messages during `npm install` and follow the instruction if the message offered any.
+- Check the package's npm page at `https://www.npmjs.com/package/<package-name>`. ie: <https://www.npmjs.com/package/express>
+  - Check the "last publish" date and versions history to get a sense of its latest activies.
+  - Open the "repository" link if it has one to view development activities.
```
```suggestion
  - Open the "repository" link if it has one to view development activities (you can run `npm repo $package` to do this as well)
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+When you've opened the package's repo, to further check development activities:
+
+- Check README or issues to see if the author indicated they are no longer interested in maintaining the package
```
```suggestion
- Check README or issues to see if the author has indicated they are no longer interested in maintaining the package
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+- Check README or issues to see if the author indicated they are no longer interested in maintaining the package
+
+- Search through existing issues and PRs to see if anyone else has submitted inquries similar to what you are seeking.
```
```suggestion
- Search through existing issues and PRs (closed and open) to see if anyone else has submitted inquiries similar to what you are seeking.
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+- Search through existing issues and PRs to see if anyone else has submitted inquries similar to what you are seeking.
+
+- Check for or file issue in the repo to encourage author to deprecate a version that:
```
```suggestion
- Check for or file an issue in the repo to encourage the author to replace and deprecate a version that:
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+- Check for or file issue in the repo to encourage author to deprecate a version that:
+
+  - Has known critical bugs and should be avoided
+  - Known to fail for LTS Node.js
```
```suggestion
  - Is known to fail for LTS node.js
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+  - It allows updating the message so you can change it later - verified as of 02/05/2019
+  - It can be undone by setting an empty message.
+    - ie: `npm deprecate package@1.0.0 ""`
+- Even if you are still maintaining the package, please consider marking an old version as deprecated.
```
I'm not so sure about this one - I deprecate old versions when they have critical bugs or vulns, but in general, using old versions is totally fine forever, and adding noise to these users' installs doesn't help anything.

jchip replied:
update to:

jchip replied:
- Please consider marking an old version with crtical issues or vulnerabilities as deprecated.

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+Other ideas for tools that could help:
+
+- A cli to allow `npm deprecate` a range of versions on a package.
```
This is already possible with npm; `npm deprecate $package@'^1||^2' $message`, say.
```suggestion
```

ljharb commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,107 @@
+
+- A cli to allow `npm deprecate` a range of versions on a package.
+- or request new support of `npm deprecate` to allow deprecating versions published before a given time
+- allow consumer to specify a specific version to use for a package through an `override` field in their package.json
```
```suggestion
 - use npm 6.9+'s aliasing ability to provide a replacement for transitive deps.
```

mcollina commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -61,7 +62,13 @@ When you've opened the package's repo, to further check development activities:
+
+- Fork the repo and enhance the code yourself.
+  - If you don't want to publish a modified copy under a different name, then consider using a tool like <https://github.com/ds300/patch-package>.
+- Consider becoming a maintainer for the package.
```
How can somebody do so? we need to add some explanation/link here. https://docs.npmjs.com/misc/disputes.html is the source of truth.

jchip replied:
The npm disputes guideline looks to be mainly for package name conflict or squatting resolution.

For a random user from the internet to take over a package, we need to establish their intent else we risk the same thing as the event-stream incident.

Then we need to establish how to gain ownership.  If the owner can at least be reached and gladly hand off ownership then that's easy.  Otherwise one could potentially use the disputes process.

I think this WG could help facilitate the process by maintaining a doc of packages that's pending seeking new maintainer and contact with the original owner.

That's why this WG would serve to help with getting ownership parking to eventually hand it off.

I think the part where "how can we establish this user's intent and let them take over this package" is one of the key things that triggered some of the discussions.

Is establishing some steps for vetting a user to take ownership part of this WG's purpose?

dominykas replied:
I think linking to the disputes doc is a good idea - it does cover the case of unmaintained packages and it does provide the steps and this group would probably have to use that process anyways - the decision to hand over the package will still fall onto npm, so even if the request goes via this WG (this should still be the last resort), the link is handy.

> Is establishing some steps for vetting a user to take ownership part of this WG's purpose?

 I think we discussed that it is. Covered in https://github.com/nodejs/package-maintenance/issues/155?

dominykas commented on docs/drafts/DEPRECATE-GUIDELINES.md:
```diff
@@ -0,0 +1,115 @@
+
+- Use the `npm deprecate` command to mark package as unmaintained.
+- Add a note to your repo's `README.md` file.
+- Reach out to the Node.js package-maintenance working group to offer us ownership so we can hand it off to a new owner.
```
Should this include updating the `support` levels in the `package.json`? That would require republishing the package, but maybe there's a way to make that easier, somehow built into the `deprecate` command?

bnb commented:
Nobody in today's meeting to discuss this specific issue/tool. Will hold off on discussion next meeting that someone is present to discuss.

mcollina commented:
@jchip are you still interested in getting this PR landed? Otherwise, is there anybody else that would like to help in moving this forward?

craftninja commented:
@jchip I'd be glad to take this over if you don't have time, no pressure!

jchip commented:
@mcollina I addressed the review comments in my last commit and I don't see any further comments.  I was hoping for enough approves to move this to be considered for merging. 

@craftninja Thanks for the offer.  I'd like to see if there're anymore comments and see if this PR can get enough approves to be considered merging.  I generally have problem attending every other meeting due to me being in PST, but will see if I can get this discussed and merge in next meeting.

jchip commented:
@ljharb thanks, going through them again.

On a side note, I miss them partly due to github has this curious way of reseting the page to the top after you commit each suggested changes, and then you have to scroll down, expand comments, and find back to where you were.  Is that just me?  Is there a better way?

jchip commented:
@ljharb that should be all of them.  Please review.  Thanks.

Eomm commented:
Only for note.

In a following PR, I think it could be worth adding some node args like [--throw-deprecation](https://nodejs.org/api/cli.html#cli_throw_deprecation)

mhdawson commented:
Looks like this is ready to land.
